lib/shared-method.js

```diff
@@ -304,86 +302,27 @@ function escapeRegex(d) {
  *
  * @param  {*} uarg            Argument value.
  * @param  {Object} desc       Argument description.
+ * @param  {Context} ctx       Remoting request context.
  * @return {*}                 Coerced argument.
  */
-function coerceAccepts(uarg, desc) {
-  var message;
+function validateInputArgument(uarg, desc, ctx) {
   var name = desc.name || desc.arg;
-  var targetType = convertToBasicRemotingType(desc.type);
-  var targetTypeIsArray = Array.isArray(targetType) && targetType.length === 1;
-
-  // If coercing an array to an erray,
-  // then coerce all members of the array too
-  if (targetTypeIsArray && Array.isArray(uarg)) {
-    return uarg.map(function(arg, ix) {
-      // when coercing array items, use only name and type,
-      // ignore all other root settings like "required"
-      return coerceAccepts(arg, {
-        name: name + '[' + ix + ']',
-        type: targetType[0],
-      });
-    });
-  }
-
-  var actualType = SharedMethod.getType(uarg, targetType);
-
-  // convert values to the correct type
-  // TODO(bajtos) Move conversions to HttpContext (and friends)
-  // SharedMethod should only check that argument values match argument types.
-  var conversionNeeded = targetType !== 'any' &&
-    actualType !== 'undefined' &&
-    actualType !== targetType;
-
-  if (conversionNeeded) {
-    // JSON.parse can throw, so catch this error.
-    try {
-      uarg = convertValueToTargetType(name, uarg, targetType);
-      actualType = SharedMethod.getType(uarg, targetType);
-    } catch (e) {
-      message = g.f('invalid value for argument \'%s\' of type ' +
-        '\'%s\': %s. Received type was %s. Error: %s',
-        name, targetType, uarg, typeof uarg, e.message);
-      throw badArgumentError(message);
-    }
-  }
-
-  var typeMismatch = targetType !== 'any' &&
-    actualType !== 'undefined' &&
-    targetType !== actualType &&
-    // In JavaScript, an array is an object too (typeof [] === 'object').
-    // However, SharedMethod.getType([]) returns 'array' instead of 'object'.
-    // We must explicitly allow assignment of an array value to an argument
-    // of type 'object'.
-    !(targetType === 'object' && actualType === 'array');
-
-  if (typeMismatch) {
-    message = g.f('Invalid value for argument \'%s\' of type ' +
-      '\'%s\': %s. Received type was converted to %s.',
-      name, targetType, uarg, typeof uarg);
-    throw badArgumentError(message);
-  }
 
   // Verify that a required argument has a value
-  // FIXME(bajtos) "null" should be treated as no value too
-  if (actualType === 'undefined') {
-    if (desc.required) {
+  if (desc.required) {
+    var argIsNotSet = uarg === null || uarg === undefined || uarg === '';
```

Comment:
The distinction between `uarg == null` and `uarg === null` is not immediately obvious and it's easy to miss when reading the code. I prefer to make this check explicit.